pallets/jobs/src/functions.rs

```diff
@@ -456,16 +462,56 @@ impl<T: Config> Pallet<T> {
 		T::MPCHandler::verify(JobWithResult {
 			job_type: job_info.job_type.clone(),
 			phase_one_job_type: Some(phase_one_job_info.job_type),
-			result: job_result,
+			result: job_result.clone(),
 		})?;
 
 		let result = PhaseResult {
 			owner: job_info.owner.clone(),
-			expiry: job_info.expiry,
+			ttl: job_info.ttl,
 			job_type: job_info.job_type.clone(),
-			result: info.proof(),
+			result: job_result,
 			permitted_caller: job_info.job_type.clone().get_permitted_caller(),
 		};
 		Ok(result)
 	}
+
+	pub fn on_idle_remove_expired_jobs(
+		now: BlockNumberFor<T>,
+		mut remaining_weight: Weight,
+	) -> Weight {
+		// early return if we dont have enough weight to perform a read
+		if remaining_weight.is_zero() {
+			return remaining_weight
+		}
+
+		// fetch all known results
+		let known_results = KnownResults::<T>::iter();
+		// we use size hint to avoid reallocations
+		let known_results_len = known_results.size_hint().0 as u64;
+		remaining_weight =
+			remaining_weight.saturating_sub(T::DbWeight::get().reads(known_results_len));
+
+		let results_to_remove = known_results.filter_map(|(role_type, job_id, result)| {
+			if result.ttl < now {
+				Some((role_type, job_id))
+			} else {
+				None
+			}
+		});
+
+		for (role_type, job_id) in results_to_remove {
+			// remaining_weight =
+			// 	remaining_weight.saturating_sub(T::DbWeight::get().reads(One::one()));
+
+			if remaining_weight.is_zero() {
+				break
+			}
+
+			Self::deposit_event(Event::<T>::JobResultExpired { role_type, job_id });
+			// DO NOT REMOVE RESULTS KEEP THEM FOR NOW.
+			// KnownResults::<T>::remove(role_type, job_id);
+		}
+
+		remaining_weight
```

Comment:
In this loop, we do emit events of jobs that are ended, and can not be used anymore, however @drewstone mentioned that we should not remove this from the storage, shall we move it somewhere else? Like `ArchivedResults`/`ExpiredResults`? do not emit events at all? 

The side effect here, is that we will always emit these events unless we do something with these expired jobs, like marking them as touched or something!

Reply:
On another consideration, it is possible to delegate this task to the gadget or worker.